ciongke/cmd/test-pr/main.go

```diff
@@ -0,0 +1,203 @@
+	}
+
+	// Set logging to write to both stdout and the build log file.
+	buildLogPath := filepath.Join(workspace, "build-log.txt")
+	file, err := os.Create(buildLogPath)
+	if err != nil {
+		// Quit here, we won't be able to upload the build log anyway.
+		log.Fatalf("Could not open log file (%s): %s", buildLogPath, err)
+	}
+	log.SetOutput(io.MultiWriter(file, os.Stdout))
+
+	startTime := time.Now()
+	result := gcs.Success
+
+	clonePath := filepath.Join(workspace, "repo")
+	commit, trusted, mergeable, err := checkoutPR(repo, clonePath, *author, *pr)
+	if trusted != nil {
+		// TODO: Status update to "Add author to .test-whitelist.txt"
+		log.Printf("Cannot proceed: %s", trusted)
+		result = gcs.Failure
+	} else if mergeable != nil {
+		// TODO: Status update to "Needs rebase"
+		log.Printf("Cannot proceed: %s", mergeable)
+		result = gcs.Failure
+	} else if err != nil {
+		log.Printf("Error preparing git: %s", err)
+		result = gcs.Failure
+	}
+
+	buildNumber, err := gcs.IncrementAndGetBuildNumber(*bucket, jobName)
+	if err != nil {
+		log.Printf("Could not get build number: %s", err)
+		result = gcs.Failure
+	}
+	buildPath := jobName + "/" + strconv.Itoa(buildNumber)
+	if err = gcs.UploadStarted(*bucket, buildPath, commit); err != nil {
+		log.Printf("Could not upload started.json: %s", err)
+		result = gcs.Failure
+	}
+
+	// Only run tests if previous steps were successful.
+	if result == gcs.Success {
+		// TODO: Status update to "Test started"
+		timeout := time.Duration(*timeoutMin) * time.Minute
+		if err = build(clonePath, timeout); err == nil {
+			// TODO: Status update to "Success"
+			log.Printf("Build succeeded after %v", time.Since(startTime))
+		} else {
+			// TODO: Status update to "Failure"
+			log.Printf("Build failed after %v: %s", time.Since(startTime), err)
+			result = gcs.Failure
+		}
+	}
+
+	if err = gcs.UploadFinished(*bucket, buildPath, result); err != nil {
+		log.Printf("Could not upload finished.json: %s", err)
+	}
+	if err = gcs.UploadBuildLog(*bucket, buildPath, buildLogPath); err != nil {
+		log.Printf("Could not upload build log: %s", err)
+	}
+}
+
+// checkoutPR returns the commit hash, an error if the author is untrusted, an
+// error if the PR can't be merged, and an error if anything else fails.
+func checkoutPR(repoPath, clonePath, prAuthor string, prNumber int) (string, error, error, error) {
+	cloneCommand := exec.Command("git", "clone", repoPath, clonePath)
+	fetchCommand := exec.Command("git", "fetch", "origin", "pull/"+strconv.Itoa(prNumber)+"/head:pr")
+	revParseCommand := exec.Command("git", "rev-parse", "pr")
+	mergeCommand := exec.Command("git", "merge", "pr", "--no-edit")
+	if err := cmdlog.RunWithLogs(cloneCommand); err != nil {
+		return "", nil, nil, err
+	}
+	// Check that the author is in master's whitelist.txt.
+	if err := validAuthor(clonePath, prAuthor); err != nil {
+		// DO NOT MERGE with this line commented out.
+		// return "", err, nil, nil
```

Comment:
Note: This makes it run tests for all authors, which I need for testing because there is no .test-whitelist.txt file in master.